File: solutions/day17.py
import heapq
from collections import defaultdict
from functools import cache
from math import inf
from operator import attrgetter

from utils.utils import split_lines


class State:
    def __init__(self, coord, direction, remaining, traversed):
        self.coord = coord
        self.direction = direction
        self.remaining = remaining
        self.traversed = traversed

    def __lt__(self, other):
        return self.traversed < other.traversed

    def __hash__(self):
        return hash((self.coord, self.direction, self.remaining))

    def __repr__(self) -> str:
        return (self.coord, self.direction, self.remaining, self.traversed).__repr__()


def parse(lines):
    return {
        complex(i, j): int(char)
        for j, line in enumerate(lines)
        for i, char in enumerate(line)
    }


@cache
def manhattan(x, y):
    return abs(x.real - y.real) + abs(x.imag - y.imag)


def A_star(graph, start, max_straight, min_turn_time):
    xmin = ymin = 0
    move_threshold = max_straight - min_turn_time
    xmax = int(max(graph.keys(), key=attrgetter("real")).real)
    ymax = int(max(graph.keys(), key=attrgetter("imag")).imag)
    # goal = complex(xmax, ymax)
    goal = complex(xmax, ymax)

    dist = defaultdict(lambda: inf)
    queue = [
        State(start, 1, max_straight, 0),
        State(
            start,
            1j,
            max_straight,
            0,
        ),
    ]
    heapq.heapify(queue)
    # Tip from Discord
    answer = max(graph.values()) * (xmax + ymax)

    def verify_coord(coord):
        return xmin <= coord.real <= xmax and ymin <= coord.imag <= ymax

    def add_neighbor(new_coord, direction, this_traversed, new_remaining):
        heapq.heappush(
            queue,
            State(
                new_coord,
                direction,
                new_remaining,
                this_traversed ,
            ),
        )

    directions = ((-1, 1), (-1j, 1j))
    while True:
        try:
            current = heapq.heappop(queue)
        except IndexError:
            break
        
        new_coord = current.coord + current.direction
        if not verify_coord(new_coord):
            continue
        this_traversed = current.traversed + graph[new_coord]
        new_remaining = current.remaining - 1
        #current_hash = hash(current)
        key = (new_coord, current.direction, new_remaining)

        if not (
            new_remaining > -1
            and this_traversed < dist[key]
            and this_traversed < answer
        ):
            continue
        dist[key] = this_traversed

        if new_coord == goal and new_remaining <= move_threshold:
            answer = this_traversed
            continue
        # Going left/right

        # print(current)
        if new_remaining <= move_threshold:
            for dir in directions[bool(current.direction.real)]:
                add_neighbor(new_coord, dir, this_traversed, max_straight)
        add_neighbor(new_coord, current.direction, this_traversed, new_remaining)

    return answer


raw_input = split_lines("inputs/day17.txt")
graph = parse(raw_input)
part1 = A_star(graph, 0, 3, 0)
print(part1)

part2 = A_star(parse(raw_input), 0, 10, 4)
print(part2)

File: solutions/day17.js
const fs = require("fs");

function parse(lines){ 
    return [...lines.map((x) => x.split("").map((x) => parseInt(x)))];
}

function pair_equal(x, y){ 
    return (x[0] == y[0]) && (x[1] == y[1]);
}

function dijkstra(start, goal, graph){
    let max_straight = 3;
    let xmin = ymin = 0;
    let xmax = graph[0].length -1;
    let dist = {};
    let part1 = Infinity;
    let ymax = graph.length -1;
    let queue = [{"x" : start[0], "y" : start[1], "direction" : [1, 0], "remaining" : 3, "distance" : graph[start[1]][start[0]]},{"x" : start[0], "y" : start[1], "direction" : [0, 1], "remaining" : 3, "distance" : graph[start[1]][start[0]]}] 

    // Only left, right, straight
    while(queue.length){ 
         let best = -Infinity;
         let choice = NaN;
         for(let i =0; i< queue.length; i++){ 
             if (queue[i]["distance"] > best){ 
                 best = queue[i]["distance"];
                 choice = i;
             }
         }
        let current = queue.pop();
        let x = current["x"];
        let y = current["y"];
        let direction = current["direction"];
        let this_distance = current["distance"];

        let key = [x, y, direction, current["remaining"]].toString();
        if ((key in dist) && this_distance >= dist[key]){ 
             continue;
         }
        dist[key] = this_distance;
        let at_goal = pair_equal([x, y], goal);
        //console.log(current)
        if(at_goal && this_distance < part1){ 
            part1 = this_distance;
        }else if (!(at_goal) && (this_distance <= part1)){ 
            let new_remaining = current["remaining"] -1;
            //Going left/right
            if (Math.abs(direction[0]) == 1){ 
                if(y > ymin){ 
                    queue.push({"x" : x, "y" : y -1, "direction" : [0, -1], "remaining" : max_straight, "distance" : this_distance + graph[y-1][x] });
                }
                if(y < ymax){ 
                    queue.push({"x" : x, "y" : y +1, "direction" : [0, 1], "remaining" : max_straight, "distance" : this_distance + graph[y + 1][x] });
                }
                // Straight
                let new_x = x + direction[0];
                if ( (new_remaining > 0) && (new_x >= xmin) && (new_x <= xmax)){ 
                    queue.push({"x" : new_x, "y" : y, "direction" : direction, "remaining" : new_remaining, "distance" : this_distance+ graph[y][new_x] })
                }
            // Going up/down
            }else{
            if(x > xmin){ 
                queue.push({"x" : x-1, "y" : y , "direction" : [-1, 0], "remaining" : max_straight, "distance" : this_distance + graph[y][x - 1] });
            }
            if(x < xmax){ 
                queue.push({"x" : x+1, "y" : y , "direction" : [1, 0], "remaining" : max_straight, "distance" : this_distance + graph[y ][x +1] });
            }
            // Straight
            let new_y = y + direction[1];
            if ( (current["remaining"] > 0) && (new_y >= ymin) && (new_y <= ymax )) { 
                queue.push({"x" : x , "y" : new_y, "direction" : direction, "remaining" : new_remaining, "distance" : this_distance+ graph[new_y][x] });
            }
            }
        }
        }
    console.log(dist)
    return part1;
}

const raw_input = fs.readFileSync('inputs/day17.txt', 'utf-8').toString().replace(/\n+$/, "").split("\n");
const graph = parse(raw_input);
const goal = [graph[0].length -1, graph.length -1];
const start = [0, 0]
const part1 = dijkstra(start, goal, graph);
console.log(part1);
